HTML parsing, socket for feedback and email notification if wanted

Molyna opened the pull request:
Need to confirm that the parsing everything freely to html is fine (assumption is that the actual ardoq application handles it)

Socket also needs checking with how much data we actually want to transmit and do we want more/less?

expez commented on resources/form.html:
```diff
@@ -92,7 +92,14 @@
+      <div class="form-group">
+        <div class="col-sm-offset-2 col-sm-10 checkbox" id="Notify">
+          <label>
+            <input name="notifier" type="checkbox"> Send you an e-mail upon completion? (Useful for large documents)
```
Tror "Email notification on completion" er bedre copy.

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -38,22 +40,27 @@
+        (do (socket-send "Valid Swagger - Importing Swagger 1" false)
             (swagger/import-swagger client spec url name headers)
             (println "Done importing swagger doc from " url "."))
         (do (println "Not a valid Swagger file\nError: ") 
```
Er det ikke bedre å bruke en logge enn å sprinkle `println` rundt omkring?

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -68,46 +75,77 @@
+      (catch Exception e
+        (println "Failed to send e-mail")))
+    )
+)
```
litt weird formatting av trailing parens her

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -68,46 +75,77 @@
+  (let [url (str (:url client) "/api/user/notify/email")]
+    (try
+      (->> {:subject "Swagger import failed" :body (str "Your Open API (Swagger) workspace is failed to import.\n" e)} 
+           (generate-string)
```
superfluous "is"

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -68,46 +75,77 @@
+             (catch com.fasterxml.jackson.core.JsonParseException e
+               (notify-failure session client "Failed to parse swagger endpoint")
+               (.printStackTrace e)
+               {:status 406
```
406 brukes for å indikere feil ved content negotiation.

f.eks hvis du ber om xml i header, server bare har json og ikke tilbyr en default representation.

400 - bad request er bedre her, fordi brukeren ber oss om å jobbe mot et endepunkt vi ikke klarer parse.

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -68,46 +75,77 @@
       (version1 client spec url wsname headers ignore-validate))))
 
-(defn swagger-api [{:keys [config]}]
+(defn notify-success [wid org session client]
```
Bedre navn her er `send-{success,failure}-email!` så det kommer frem at vi 1) gjør IO 2) Faktisk sender en mail og ikke f.eks sender melding over ws.

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -68,46 +75,77 @@
+             (catch IllegalArgumentException e
+               (notify-failure session client "Swagger file does not conform to Swagger specification")
+               (.printStackTrace e)
                {:status 406
```
400 er også bedre statuskode her

expez commented on src/ardoq/swagger/api.clj:
```diff
@@ -68,46 +75,77 @@
+             (catch clojure.lang.ExceptionInfo e
+               (notify-failure session client "An unexpected error occured!")
+               (.printStackTrace e)
+               (if (= 404 (-> e ex-data :status))
```
404 (not found) brukes bare når klienten ber om en ressurs som serveren ikke finner / kjenner til. 

statuskode 500 (internal server error) brukes for "Noe gikk galt på serveren.  Du får ikke vite hva :p"

Forøvrig finnes det en supernice emacs pakke som heter [know your http well](https://github.com/for-GET/know-your-http-well) som lar deg slå opp hva statuskoder, headers og alt mulig HTTP relatert betyr.  @ebaxt kan sikkert også like den

expez commented on src/ardoq/swagger/swagger.clj:
```diff
@@ -257,7 +261,9 @@
              :operations operations
              :refs refs}]
     (delete-resources client workspace (:apis resource-listing))
+    (socket-send "Removed surplus resources")
```
Tror jeg liker "Done pruning resources" og pruning operations under

expez commented:
Aight dette ser bra ut!

Tror det vil være bedre hvis vi forteller hva vi holder på med istedet for hva vi har gjort.  Så i stedet for i loggen:

```
Created 5 models
Pruned 9 resources
Glorious victory!  Update complete!
```

Så står det 

```
Creating models...
Pruning resources...
Winning! All done!
```

expez commented:
Funket ikke helt som det skulle.  Mangler secure websocket connection:

![image](https://cloud.githubusercontent.com/assets/1006557/13846522/f8c54a5e-ec47-11e5-8e32-39e39b438bfe.png)

Molyna commented:
Updated according to comments. Fixed the minor spelling errors, added a check for secure vs insecure when connecting to the socket and set better http status codes (we now use 400, 404, 422 and 500 for error messages).

The 404 is staying, it is an actual 404. Swagger 1 had a setup where you had the main file, this then pointed to a new document on a new url. If these extra urls throw a 404 we check and then throw our own 404 as a result. 

The 422 is in accordance with what people are doing now for valid but semantically incorrect  content. And in case the swagger file isn't validated I feel that fits with it better than the 400 (it also allows for some different error codes depending on situations). 

I also updated swagger 1 in accordance with how we do swagger 2 in the socket. So the status messages go this way

```
Updating models
Updated 22 models
Updating resources
Updated 30 resources
```

It also tries to report how many it is updating each time, although this isn't possible always without going very spammy (like with swagger 1).

Molyna commented:
Er denne testet igjen, @expez?

expez commented:
@Molyna Nei, har ikke gjort det enda.

ebaxt commented:
@Molyna Still doesn't use wss when called over https it looks like.
![screen shot 2016-03-31 at 18 04 25](https://cloud.githubusercontent.com/assets/76092/14182448/864ea2ac-f76b-11e5-99ae-6855ff02dc5c.png)

Molyna commented:
Latest commit should correctly extract the protocol type and create a socket accordingly.

Molyna commented:
I'll test this properly myself to make sure we don't get things like yesterday before this pull request goes through.
